arch: riscv: Add function to clear all unlocked PMP entries

fsammoura1980 opened the pull request:
Introduce the new function `riscv_pmp_clear_all()` to securely reset the Physical Memory Protection (PMP) configuration.

This function iterates through all configured PMP slots. For each entry, it checks if the Lock (L) bit is set. If the entry is not locked, it clears the Address Matching Mode (A) bits, effectively setting the region type to OFF (Null Region), disabling the entry.

The function ensures it operates in default Machine Status Registers (CSRs) with `MSTATUS.MPRV = 0` and `MSTATUS.MPP = 11`.

This provides a mechanism to clear all non-locked PMP regions, returning them to a default disabled state. The function is exposed in the pmp.h header file.

jimmyzhe commented:
According to the [The RISC-V Instruction Set Manual: Volume II: Privileged Architecture
](https://github.com/riscv/riscv-isa-manual/blob/main/src/machine.adoc#memory-privilege-in-mstatus-register), the mstatus.MPRV and mstatus.MPP bits affect load/store behavior, while PMP registers are CSRs.

CSR is accessed via csrr/csrw ... instructions, not load/store instructions, and the permission depend on the current privilege level, not on mstatus.MPRV or mstatus.MPP.

I think this setup is not related to permission for access PMP CSR. Is this setup needed?

fsammoura1980 committed:
arch: riscv: Add function to clear all unlocked PMP entries

Introduce the new function `riscv_pmp_clear_all()` to securely reset
the Physical Memory Protection (PMP) configuration.

This function iterates through all configured PMP slots. For each entry,
it checks if the Lock (L) bit is set. If the entry is not locked,
it clears the Address Matching Mode (A) bits, effectively setting the
region type to OFF (Null Region), disabling the entry.

The function ensures it operates in Machine mode with MSTATUS.MPRV = 0
before modifying any PMP Control and Status Registers (CSRs).

This provides a mechanism to clear all non-locked PMP regions,
returning them to a default disabled state. The function is exposed
in the pmp.h header file.

Signed-off-by: Firas Sammoura <fsammoura@google.com>
